File: contests/itmo/masters-3-sem/005/C_Predecessors-In-Tree.cpp
#include <iostream>
#include <vector>
#include <string>
#include <utility>

class Graph {
    std::vector<std::vector<int>> adj;
    std::vector<bool> visited;
    std::vector<int> entryTime;
    std::vector<int> exitTime;
    int timer;

public:
    Graph(int n) : adj(n), visited(n), entryTime(n), exitTime(n), timer(0) {}

    void add(int u, int v) {
        adj[u].push_back(v);
    }

    void dfs(int v) {
        visited[v] = true;
        timer++;
        entryTime[v] = timer;

        for (int u : adj[v]) {
            if (!visited[u]) {
                dfs(u);
            }
        }

        timer++;
        exitTime[v] = timer;
    }

    bool isPredecessor(int u, int v) {
        return entryTime[u] <= entryTime[v] && exitTime[u] >= exitTime[v];
    }
};

int main() {
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int n, m;
    std::cin >> n >> m;

    Graph g(n);

    if (n > 1) {
        for (int i = 0; i < n - 1; i++) {
            int parent;
            std::cin >> parent;
            g.add(parent - 1, i + 1);
        }
    }

    std::vector<std::pair<int, int>> pairs(m);
    for (int i = 0; i < m; i++) {
        int u, w;
        std::cin >> u >> w;
        pairs[i] = std::pair(u - 1, w - 1);
    }

    g.dfs(0);

    for (const auto& p : pairs) {
        std::cout << (g.isPredecessor(p.first, p.second) ? 1 : 0) << '\n';
    }

    return 0;
}

File: contests/itmo/masters-3-sem/005/C_Predecessors-In-Tree.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Pair struct {
	u int
	w int
}

type Graph struct {
	adj       [][]int
	visited   []bool
	entryTime []int
	exitTime  []int
	timer     int
}

func newGraph(n int) *Graph {
	return &Graph{
		adj:       make([][]int, n),
		visited:   make([]bool, n),
		entryTime: make([]int, n),
		exitTime:  make([]int, n),
	}
}

func (g *Graph) add(u, v int) {
	g.adj[u] = append(g.adj[u], v)
}

func (g *Graph) dfs(v int) {
	g.visited[v] = true
	g.timer++
	g.entryTime[v] = g.timer

	for _, u := range g.adj[v] {
		if !g.visited[u] {
			g.dfs(u)
		}
	}

	g.timer++
	g.exitTime[v] = g.timer
}

func (g *Graph) isPredecessor(u, v int) bool {
	return g.entryTime[u] <= g.entryTime[v] && g.exitTime[u] >= g.exitTime[v]
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)

	scanner.Scan()
	parts := strings.Fields(scanner.Text())
	n, _ := strconv.Atoi(parts[0])
	m, _ := strconv.Atoi(parts[1])

	g := newGraph(n)

	scanner.Scan()
	if n > 1 {
		parents := strings.Fields(scanner.Text())
		for i := 0; i < n-1; i++ {
			parent, _ := strconv.Atoi(parents[i])
			g.add(parent-1, i+1)
		}
	}

	pairs := make([]Pair, m)
	for i := 0; i < m; i++ {
		scanner.Scan()
		parts = strings.Fields(scanner.Text())
		u, _ := strconv.Atoi(parts[0])
		w, _ := strconv.Atoi(parts[1])
		pairs[i] = Pair{u: u - 1, w: w - 1}
	}

	g.dfs(0)

	for _, p := range pairs {
		if g.isPredecessor(p.u, p.w) {
			fmt.Println(1)
		} else {
			fmt.Println(0)
		}
	}
}
